### PR TITLE
[Kernel][Hardware][Amd]Custom paged attention kernel for rocm

### PR DESCRIPTION
**Pull request by charlifu**
This PR adds custom paged attention kernel support for rocm.

- add a separate folder for rocm kernel, which enables more specialized kernels for rocm in the future.
- add custom paged attention for rocm
- enable rocm custom paged attention kernel for specific size and dtype.
- add unit test

Todo:
- add fp8 kv cache support for rocm custom paged attention.
- add navi support for rocm custom paged attention.

Performance improvement (llama-2-70b-chat-hf):

| num-prompt | input-len | output-len | w/o             | w/               |
|------------|-----------|------------|-----------------|------------------|
| 8          | 512       | 256        | 360.15 tokens/s | 419.62 tokens/s  |
| 8          | 512       | 512        | 271.54 tokens/s | 293.50 tokens/s  |
| 16         | 512       | 256        | 616.33 tokens/s | 770.72 tokens/s  |
| 16         | 512       | 512        | 414.81 tokens/s | 560.03 tokens/s  |
| 32         | 512       | 256        | 913.00 tokens/s | 1263.37 tokens/s |
| 32         | 512       | 512        | 626.82 tokens/s | 969.73 tokens/s  |

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Review by WoosukKwon — COMMENTED** [review]
@charlifu Thanks for upstreaming the kernel and also adding the test!

It looks good to me overall, but I have a few suggestions:
1. Move `use_rocm_paged_attention` in `selected.py` to `rocm_flash_attn.py`.
2. Cache `rocm_paged_attention_available` in `use_rocm_paged_attention` to minimize its overheads.
3. In `rocm_flash_attn.py`, use `rocm_paged_attention_available` to decide between `PagedAttention.forward_decode` and `ops.paged_attention_rocm`. This way, we will be able to keep `paged_attn.py` untouched by this PR.

File: vllm/envs.py [hunk 1]
```diff
@@ -203,6 +203,11 @@ def get_default_config_root():
     (os.environ.get("VLLM_DYNAMO_USE_CUSTOM_DISPATCHER", "True").lower() in
      ("true", "1")),
 
+    # Rocm custom paged attention implemented for MI3* GPUs
+    "VLLM_USE_ROCM_CUSTOM_PAGED_ATTN":
+    lambda: (os.getenv("VLLM_USE_ROCM_CUSTOM_PAGED_ATTN", "True").lower() in
+             ("true", "1") != "0"),
```
[the remark this answers is not in the document]
Comment (charlifu) [reply vllm/envs.py:209]: will remove this.

File: tests/kernels/test_attention.py [hunk 2]
```diff
@@ -312,13 +314,173 @@ def ref_multi_query_kv_attention(
     return torch.cat(ref_outputs, dim=0)
 
 
+@pytest.mark.parametrize("version", ["rocm"])
+@pytest.mark.parametrize("num_seqs", NUM_GEN_SEQS)
+@pytest.mark.parametrize("num_heads", NUM_HEADS)
+@pytest.mark.parametrize("head_size", [64, 128])  # only test 64 128
```
[the remark this answers is not in the document]
Comment (charlifu) [reply tests/kernels/test_attention.py:320]: Other sizes are not supported yet, we will add them in subsequent PR.

File: vllm/_custom_ops.py [hunk 3]
```diff
@@ -17,6 +17,10 @@
     except ImportError as e:
         logger.warning("Failed to import from vllm._C with %r", e)
 
+if current_platform.is_rocm():
+    with contextlib.suppress(ImportError):
```
[the remark this answers is not in the document]
Comment (charlifu) [reply vllm/_custom_ops.py]: Done

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for addressing my review!
